File: 2-2/전공/알고리즘/과제/코딩과제1-정렬알고리즘/bubbleSort/bubblesort.cpp
#include <iostream> 
using namespace std;

#define swap(x,y,t) ((t)=(x),(x)=(y),(y)=(t))

void bubbleSort(A, m);
void bubbleSort_imp1(A, n);
void bubbleSort_imp2(A, n);
int main()
{
    int N;
    cin >> N;
    while (N--)
    {
        int M;
        cin >> M;
        arr1[1000] = { 0 };
        arr2[1000] = { 0 };
        arr3[1000] = { 0 };
        A = { 0,0 };
        B = { 0,0 };
        C = { 0,0 };
        for (int i = 0; i < M; i++) {
            int n;
            cin >> n;
            arr1[i] = n;
            arr2[i] = n;
            arr3[i] = n;
        }
        bubbleSort(arr, M);
        bubbleSort_imp1(arr, M);
        bubbleSort_imp2(arr, M);
        cout << A[0] << " " << A[1] << " " << B[0] << " " << B[1] << " " << C[0] << " " << C[1]<<" "<<"\n";
    }
        
            
}
void bubbleSort(A, m)
{
    int compare_cnt = 0;
    int swap_cnt = 0;
    for (pass = 1; pass < m; pass++)
        for (i = 1; i <= m - pass; i++)
        {
            compare_cnt++;
            if (A[i - 1] > A[i]) 
            {
                swap_cnt++;
                swap(A[i - 1], A[i]);
            }
        }
    A[0] = compare_cnt;
    A[1] = swap_cnt;
}

void bubbleSort_imp1(A, n)
{
    int compare_cnt = 0;
    int swap_cnt = 0;

    for (pass = 1; pass < n; pass++) 
    {
        bool swapped = false; 
        for (i = 1; i <= n - pass; i++)
            compare_cnt++;
            if (a[i - 1] > a[i]) 
            {
                swap(a[i - 1], a[i]);
                swap_cnt++;
                swapped = true;
            }
        if (swapped == false) 
            break;
    }
    B[0] = compare_cnt;
    B[1] = swap_cnt;
}
    

void bubbleSort_imp2(A, n)
{
    int swap_cnt = 0;
    int compare_cnt = 0;
    int lastSwappedPos = n;
    for (lastSwappedPos > 0)
    {
        int swappedPos = 0;
        for (i = 1; i < lastSwappedPos; i++)
            compare_cnt++;
            if (a[i - 1] > a[i])
            {
                swap_cnt++;
                swap(a[i - 1], a[i]);
                swappedPos = i;
            }
        lastSwappedPos = swappedPos;
    }
    C[0] = compare_cnt;
    c[1] = swap_cnt;
}

File: 2-2/전공/알고리즘/과제/코딩과제1-정렬알고리즘/bubbleSort/bubblesort.py
import sys
def bubbleSort(A,m):
    arr = A.copy()
    swap_cnt = 0
    compare_cnt = 0
    for p in range(1,m):
        for i in range(1,m-p+1):
            compare_cnt+=1
            if(arr[i-1]>arr[i]):
                k = arr[i-1]
                arr[i-1] = arr[i]
                arr[i]=k
                swap_cnt+=1

    return[compare_cnt, swap_cnt]

def bubbleSort_imp1(A, m):
    arr = A.copy()
    swap_cnt = 0
    compare_cnt = 0
    for p in range(1,m):
        swapped = False
        for i in range(1,m-p+1):
            compare_cnt+=1
            if(arr[i-1]>arr[i]):
                k = arr[i-1]
                arr[i-1] = arr[i]
                arr[i]=k
                swapped= True
                swap_cnt+=1
        if(swapped == False):
            break;
    return[compare_cnt, swap_cnt]

def bubbleSort_imp2(A , m):
    arr = A.copy()
    swap_cnt = 0
    compare_cnt = 0
    lastswappedpos = m
    while(lastswappedpos > 0):
        swappedpos = 0
        for i in range(1,lastswappedpos):
            compare_cnt+=1
            if(arr[i-1]>arr[i]):
                k = arr[i-1]
                arr[i-1] = arr[i]
                arr[i]=k
                swappedpos = i
                swap_cnt+=1
        lastswappedpos = swappedpos
    return[compare_cnt, swap_cnt]






n = int(sys.stdin.readline())

for i in range(n):
    arr = list(map(int, sys.stdin.readline().split()))
    m = arr[0]
    arr = arr[1:]
    A = bubbleSort(arr, m)
    B = bubbleSort_imp1(arr, m)
    C = bubbleSort_imp2(arr, m)
    print(A[0]," ",A[1]," ",B[0]," ", B[1]," ",C[0]," ",C[1])
